src/parser/tests/parse_errors/parse_error_statements_tests.rs

```diff
@@ -0,0 +1,976 @@
+                    name: "c".into(),
+                    data_type: DataTypeDeclaration::DataTypeReference {
+                        referenced_type: "INT".into(),
+                    },
+                    initializer: None,
+                    location: SourceRange::undefined(),
+                },],
+                variable_block_type: VariableBlockType::Local,
+            }
+        )
+    );
+
+    assert_eq!(
+        diagnostics[0],
+        Diagnostic::unexpected_token_found(
+            format!("{:?}", Token::KeywordEndVar),
+            "'4 : INT;'".into(),
+            SourceRange::new(77..85)
+        )
+    );
+}
+
+#[test]
+fn invalid_variable_data_type_error_recovery() {
+    let lexer = lex("
+        PROGRAM p
+            VAR 
+                a DINT : ;
+                c : INT;
+            END_VAR
+        END_PROGRAM
+        ");
+
+    let (cu, diagnostics) = parse(lexer).unwrap();
+    let pou = &cu.units[0];
+    assert_eq!(
+        format!("{:#?}", pou.variable_blocks[0]),
+        r#"VariableBlock {
+    variables: [
+        Variable {
+            name: "a",
+            data_type: DataTypeReference {
+                referenced_type: "DINT",
+            },
+        },
+        Variable {
+            name: "c",
+            data_type: DataTypeReference {
+                referenced_type: "INT",
+            },
+        },
+    ],
+    variable_block_type: Local,
+}"#
+    );
+
+    assert_eq!(
+        diagnostics,
+        vec![
+            Diagnostic::missing_token("KeywordColon".into(), SourceRange::new(54..58)),
+            Diagnostic::unexpected_token_found(
+                "KeywordSemicolon".into(),
+                "':'".into(),
+                SourceRange::new(59..60)
+            )
+        ]
+    );
+}
+
+#[test]
+fn test_if_with_missing_semicolon_in_body() {
+    //regress, this used to end in an endless loop
+    let lexer = lex("PROGRAM My_PRG
+            IF TRUE THEN
+                x := x + 1
+            END_IF
+        END_PROGRAM
+    ");
+    let (_, diagnostics) = parse(lexer).unwrap();
+
+    assert_eq!(
+        diagnostics,
+        vec![
+            Diagnostic::missing_token("[KeywordSemicolon, KeywordColon]".into(), (79..85).into()),
+            Diagnostic::unexpected_token_found(
+                "KeywordSemicolon".into(),
+                "'END_IF'".into(),
+                (79..85).into()
+            )
+        ]
+    );
+}
+
+#[test]
+fn test_nested_if_with_missing_end_if() {
+    //regress, this used to end in an endless loop
+    let lexer = lex("PROGRAM My_PRG
+            IF FALSE THEN
+                IF TRUE THEN
+                    x := y;
+            END_IF
+            y := x;
+        END_PROGRAM
+    ");
+    let (unit, diagnostics) = parse(lexer).unwrap();
+
+    assert_eq!(
+        diagnostics,
+        vec![
+            Diagnostic::missing_token(
+                "[KeywordElseIf, KeywordElse, KeywordEndIf]".into(),
```

Comment:
In the endif check of the parser, make endif the first option for better errors

Reply:
done